deployer/src/lib.rs

```diff
@@ -1,4 +1,604 @@
+            }))
+        );
+        let service = match self.dal.service(&deployment.service_id).await {
+            Ok(inner) => inner,
+            Err(err) => {
+                return Ok(Response::new(DestroyDeploymentResponse {
+                    success: false,
+                    message: Some(format!("error fetching service from persistence: {}", err)),
+                }))
+            }
+        };
+
+        if service.state_variant != ServiceRunning::name() {
+            return Ok(Response::new(DestroyDeploymentResponse {
+                success: false,
+                message: Some("the deployment is not running".to_string()),
+            }));
+        }
+
+        // Destroying the deployment and waiting on finishing up
+        let dal = self.dal.clone();
+        let task_router = self.task_router.clone();
+        let docker = self.docker.clone();
+        let runtime_manager = self.runtime_manager.clone();
+        let sender = self.deployment_state_machine_channel.clone();
+
+        // Destroy the existing service sandbox.
+        match TaskBuilder::new(dal)
+            .task_router(task_router)
+            .service_id(deployment.service_id)
+            .service_docker_context(ServiceDockerContext::new(docker, runtime_manager))
+            .and_then(task::destroy())
+            .and_then(task::run_until_done())
+            .send(&sender)
+            .await
+        {
+            Ok(handle) => handle.await,
+            Err(err) => {
+                return Ok(Response::new(DestroyDeploymentResponse {
+                    success: false,
+                    message: Some(err.to_string()),
+                }));
+            }
+        };
+
+        Ok(Response::new(DestroyDeploymentResponse {
+            success: true,
+            message: None,
+        }))
+    }
+
+    #[instrument(skip_all)]
+    async fn target_ip(
+        &self,
+        request: tonic::Request<TargetIpRequest>,
+    ) -> TonicResult<tonic::Response<TargetIpResponse>, tonic::Status> {
+        // Authorize the request.
+        request.verify(Scope::ServiceRead)?;
+        let claim = request.extensions().get::<Claim>().cloned();
+        let request = request.into_inner();
+
+        // Do a cleanup in terms of previous invalid deployments.
+        let service_id = Ulid::from_string(&request.service_id)
+            .map_err(|_| tonic::Status::invalid_argument("invalid deployment id"))?;
+        let service = self
+            .dal
+            .service(&service_id)
+            .await
+            .map_err(|err| tonic::Status::not_found(err.to_string()))?;
+
+        if service.state.is_completed() || service.state.is_destroyed() {
+            return Ok(Response::new(TargetIpResponse {
+                success: false,
+                target_ip: None,
+                message: Some("the service is not running".to_string()),
+            }));
+        }
+
+        if service.state.is_stopped() && request.instate {
```

Comment:
Could we simplify this by just returning a target IP or a tonic::Status with a message (rather then the response with success, ip and message)? Then we can use `?`